Enable EmbeddedKafkaRest with RBAC on kafka1 (#239)
confluentinc/cp-demo: docs/index.rst

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v3/clusters | jq ".data[0].cluster_id"")

Reviewer: IIUC this won't work due to KREST-19?

Reviewer: @rnpridgeon can you please revert so we can merge?

Author: alternatively you could use the kafka rest credentials, for now its reverted.

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v3/clusters | jq ".data[0].cluster_id"")

Reviewer: @rnpridgeon cp-demo code continues to refer to the `v1/metadata/id` endpoint instead of `v3/clusters` endpoint.

1. For consistency, it should all be the same.  Need to decide on one or the other.
2. Which one should it be in 6.0.x?  What's the difference between the two endpoints, when should a user hit one versus the other?

Author: ERP is on by default, MDS isn't. So I'd personally use the new one. I thought I got them all but alas I guess I missed some.

Reviewer: 1. `http://localhost:8091/v1/metadata/id` is also used in the scripts themselves, so those would need to change as well.
2. What are the other consequences of changing the endpoint?  What if the embedded RP dies?

Author: 1. okay I'm impartial but if we want to demonstrate the new API we should use the new one
2. It goes away; this is true for the former UI if MDS dies though so using the former implementation isn't more robust.

Reviewer: Ok, in that case, please update all instances (see scripts):

```
scripts/validate/validate_rest_proxy.sh:KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
scripts/validate/validate_bindings.sh:KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
scripts/helper/functions.sh:  KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")
```

>      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v3/clusters | jq ".data[0].cluster_id"")

Reviewer: IIUC this won't work due to KREST-19?

Reviewer: @rnpridgeon can you please revert so we can merge?

Author: seems my local repo and the remote repo diverged causing some funkyness. Did a hard reset on the remote repo and reverted these again. Seems the commit suggested change button maybe more of an annoyance than just commiting the suggestion manually.

--------------------

The `Confluent REST Proxy <https://docs.confluent.io/current/kafka-rest/docs/index.html>`__  is running for optional client access.

> It may run in standalone mode or in embedded most hosted on the broker.

Reviewer: ```suggestion
This demo showcases |crest| in two modes: (1) running as a standalone service and (2) embedded on the |ak| brokers.
```

> While in embedded mode only the .. _V3 APIs: http://staging-docs.confluent.io.s3-website-us-west-2.amazonaws.com/current/kafka-rest/api.html#crest-api-v3 are supported this time.

Reviewer: ```suggestion
While in embedded mode only the :ref:`rest-proxy-v3` are supported this time.
```

> Additionally each API's URL must include the prefix ``/kafka``

Reviewer: Delete this line

> #. Use the |crest|, which is listening for HTTPS on port 8086, to try to produce a message to the topic ``users``, referencing schema id ``7``. This schema was registered in |sr| in the previous section. It should fail due to an authorization error.

Reviewer: ```suggestion
#. Use the standalone |crest|, which is listening for HTTPS on port 8086, to try to produce a message to the topic ``users``, referencing schema id ``7``. This schema was registered in |sr| in the previous section. It should fail due to an authorization error.
```

--------------------

The `Confluent REST Proxy <https://docs.confluent.io/current/kafka-rest/docs/index.html>`__  is running for optional client access.

> This demo showcases |crest| in two modes: (1) running as a standalone service and (2) embedded on the |ak| brokers.

Reviewer: ```suggestion
This demo showcases |crest| in two modes: (1) running as a standalone service listening for HTTPS requests on port 8086,  and (2) embedded on the |ak| brokers listening for HTTP requests on port 8091 on ``kafka1`` and on port 8092 on ``kafka2`` (the port is shared with the MDS listener)
```

> While in embedded mode |crest| listens for requests on port 8092 providing only the :ref:`rest-proxy-v3` are supported this time.

Reviewer: There are two ERP instances, one on broker1 and one on broker2.  In Docker they shouldn't both be listening on port 8092.  Can it be differentiated?

Author: MDS and KafkaRest share the same HTTP server so they have the same listeners. I can add the full address for both

> While in embedded mode |crest| listens for requests on http://kafka1:8091/kafka and http://kafka2:8092/kafka providing only the :ref:`rest-proxy-v3` are supported this time.

Reviewer: Delete this line

> #. Use the standalone |crest|, which is listening for HTTPS on port 8086, to try to produce a message to the topic ``users``, referencing schema id ``7``. This schema was registered in |sr| in the previous section. It should fail due to an authorization error.

Reviewer: ```suggestion
#. Use the standalone |crest| to try to produce a message to the topic ``users``, referencing schema id ``7``. This schema was registered in |sr| in the previous section. It should fail due to an authorization error.
```

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>       KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")

Reviewer: ```suggestion
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
```

From the host machine, this should be `localhost`, not `kafka1`

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>       KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")

Reviewer: ```suggestion
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
```

From the host machine, this should be `localhost`, not `kafka1`

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>       KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")

Reviewer: ```suggestion
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
```

From the host machine, this should be `localhost`, not `kafka1`

       docker-compose exec restproxy curl -X DELETE -H "Content-Type: application/vnd.kafka.v2+json" --cert /etc/kafka/secrets/restproxy.certificate.pem --key /etc/kafka/secrets/restproxy.key --tlsv1.2 --cacert /etc/kafka/secrets/snakeoil-ca-1.crt -u appSA:appSA https://restproxy:8086/consumers/my_avro_consumer/instances/my_consumer_instance

#. For the next few steps, use the |crest| that is embedded on the |ak| brokers. Only :ref:`rest-proxy-v3` is supported this time.  Create a role binding for the client to be granted ``ResourceOwner`` role for the topic ``dev_users``.

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>       KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")

Reviewer: ```suggestion
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
```

From the host machine, this should be `localhost`, not `kafka1`

      # Then create the role binding for the topic ``dev_users``
      docker-compose exec tools bash -c "confluent iam rolebinding create \
          --principal User:appSA \
          --role ResourceOwner \

>           --resource Topic:usersq \

Reviewer: ```suggestion
          --resource Topic:dev_users \
```

          --kafka-cluster-id $KAFKA_CLUSTER_ID"

#. Create the topic ``dev_users`` with embedded |crest|.

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>       KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")

Reviewer: ```suggestion
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
```

From the host machine, this should be `localhost`, not `kafka1`

>       docker-compose exec restproxy curl -X POST -H "Content-Type: application/json" -H "accept: application/json" -u appSA:appS "https://kafka1:8091/kafka/v3/clusters/${KAFKA_CLUSTER_ID}/topics" -d "{\"topic_name\":\"dev_users\",\"partitions_count\":64,\"replication_factor\":2,\"configs\":[{\"name\":\"cleanup.policy\",\"value\":\"compact\"},{\"name\":\"compression.type\",\"value\":\"gzip\"}]}"

Reviewer: Results in error:

```
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
```

>       docker-compose exec restproxy curl -X POST -H "Content-Type: application/json" -H "accept: application/json" -u appSA:appSA "http://kafka1:8091/kafka/v3/clusters/${KAFKA_CLUSTER_ID}/topics" -d "{\"topic_name\":\"dev_users\",\"partitions_count\":64,\"replication_factor\":2,\"configs\":[{\"name\":\"cleanup.policy\",\"value\":\"compact\"},{\"name\":\"compression.type\",\"value\":\"gzip\"}]}"

Reviewer: ```suggestion
      docker-compose exec restproxy curl -X POST -H "Content-Type: application/json" -H "accept: application/json" -u appSA:appSA "http://kafka1:8091/kafka/v3/clusters/${KAFKA_CLUSTER_ID}/topics" -d "{\"topic_name\":\"dev_users\",\"partitions_count\":64,\"replication_factor\":2,\"configs\":[{\"name\":\"cleanup.policy\",\"value\":\"compact\"},{\"name\":\"compression.type\",\"value\":\"gzip\"}]}" | jq
```

#. List topics with embedded |crest| to find the newly created ``dev_users``.

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID

>       KAFKA_CLUSTER_ID=$(curl -s http://kafka1:8091/v1/metadata/id | jq -r ".id")

Author: ```suggestion
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v1/metadata/id | jq -r ".id")
```

>       docker-compose exec restproxy curl -X POST -H "Content-Type: application/json" -H "accept: application/json" -u appSA:appS http://kafka1:8091/kafka/v3/clusters/${KAFKA_CLUSTER_ID}/topics

Reviewer: Results in error:

```
{
"servlet":"default",
"message":"Unauthorized",
"url":"/kafka/v3/clusters/$KAFKA_CLUSTER_ID/topics",
"status":"401"
}          
```

Note: `appS` --> `appSA`

>       docker-compose exec restproxy curl -X GET -H "Content-Type: application/json" -H "accept: application/json" -u appSA:appSA http://kafka1:8091/kafka/v3/clusters/${KAFKA_CLUSTER_ID}/topics

Reviewer: ```suggestion
      docker-compose exec restproxy curl -X GET -H "Content-Type: application/json" -H "accept: application/json" -u appSA:appSA http://kafka1:8091/kafka/v3/clusters/${KAFKA_CLUSTER_ID}/topics |jq '.data[].topic_name'
```

       docker-compose exec restproxy curl -X DELETE -H "Content-Type: application/vnd.kafka.v2+json" --cert /etc/kafka/secrets/restproxy.certificate.pem --key /etc/kafka/secrets/restproxy.key --tlsv1.2 --cacert /etc/kafka/secrets/snakeoil-ca-1.crt -u appSA:appSA https://restproxy:8086/consumers/my_avro_consumer/instances/my_consumer_instance

> #. Now you will use the |crest| that is embedded on the |ak| brokers.  Create a role binding for the client to be granted ``ResourceOwner`` role for the topic ``dev_users``.

Reviewer: ```suggestion
#. For the next few steps, use the |crest| that is embedded on the |ak| brokers. Only :ref:`rest-proxy-v3` is supported this time.  Create a role binding for the client to be granted ``ResourceOwner`` role for the topic ``dev_users``.
```

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v3/clusters | jq ".data[0].cluster_id")

      # Then create the role binding for the topic ``dev_users``
      docker-compose exec tools bash -c "confluent iam rolebinding create \
          --principal User:appSA \
          --role ResourceOwner \
          --resource Topic:usersq \
          --kafka-cluster-id $KAFKA_CLUSTER_ID"

#. Create the topic ``dev_users`` with the embedded |crest|.

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v3/clusters | jq ".data[0].cluster_id")

>       docker-compose exec restproxy curl -X POST -H "Content-Type: application/json" -H "accept: application/json" --cert /etc/kafka/secrets/restproxy.certificate.pem --key /etc/kafka/secrets/restproxy.key --tlsv1.2 --cacert /etc/kafka/secrets/snakeoil-ca-1.crt -u appSA:appS "http://localhost:8082/v3/clusters/${KAFKA_CLUSTER_ID}/topics" -d "{\"topic_name\":\"dev_users\",\"partitions_count\":64,\"replication_factor\":2,\"configs\":[{\"name\":\"cleanup.policy\",\"value\":\"compact\"},{\"name\":\"compression.type\",\"value\":\"gzip\"}]}"

Reviewer: This is hitting port 8082, isn't that contradictory with the earlier statement:

```
While in embedded mode |crest| listens for requests on http://kafka1:8091 and http://kafka2:8092 providing only the :ref:`rest-proxy-v3` are supported this time.
```

Author: Right you are. Does it just make sense to use the v3 APIs on the standalone leaving it up to the reader to try v3 apis against the embedded instance at their leisure.

Reviewer: IIUC v3 doesn't support produce/consume, so the standalone can't use v3?  If that is correct, then I would suggest:

a) standalone uses v1 with produce/consume -- requires no change to existing docs
b) ERP uses v3 with just topic creation

Reviewer 2: a) standalone uses V2 with produce/consume (V1 is gone in 6.0)

Reviewer: @rigelbm correction, cp-demo example is already using v2 for the standalone (not v1).  If 6.0 supports v2 and v3, then this PR can optionally update all the commands to use v3 syntax for standalone.

Still open issue is reconciling the ports: why is the ERP example using 8082 when the added writing says 8091 and 8092

Author: It's a typo again my suggestion is to not mix the two in the examples but instead mention that ERP listens on 809[1-2] in the intro if you want to execute v3 requests against it.

Reviewer: My strong leaning is that if this PR adds ERP to the configuration, then the PR should also add an exercise to execute statements to the ERP.

Author: Both examples point to kafka1:8091 now

       docker-compose exec restproxy curl -X DELETE -H "Content-Type: application/vnd.kafka.v2+json" --cert /etc/kafka/secrets/restproxy.certificate.pem --key /etc/kafka/secrets/restproxy.key --tlsv1.2 --cacert /etc/kafka/secrets/snakeoil-ca-1.crt -u appSA:appSA https://restproxy:8086/consumers/my_avro_consumer/instances/my_consumer_instance

> #. Create a role binding for the client providing it total ownership of topic ``dev_users``.

Reviewer: ```suggestion
#. Now you will use the |crest| that is embedded on the |ak| brokers.  Create a role binding for the client to be granted ``ResourceOwner`` role for the topic ``dev_users``.
```

   .. sourcecode:: bash

      # First get the KAFKA_CLUSTER_ID
      KAFKA_CLUSTER_ID=$(curl -s http://localhost:8091/v3/clusters | jq ".data[0].cluster_id"")

      # Then create the role binding for the topic ``dev_users``
      docker-compose exec tools bash -c "confluent iam rolebinding create \
          --principal User:appSA \
          --role ResourceOwner \
          --resource Topic:usersq \
          --kafka-cluster-id $KAFKA_CLUSTER_ID"

> #. Create the topic ``dev_users``.

Reviewer: ```suggestion
#. Create the topic ``dev_users`` with the embedded |crest|.
```
